mos-platform/pce/libpce/src/memory.S

```diff
@@ -0,0 +1,124 @@
+; Copyright (c) 2023 Adrian "asie" Siekierka
+;
+; Licensed under the Apache License, Version 2.0 with LLVM Exceptions,                                   
+; See https://github.com/llvm-mos/llvm-mos-sdk/blob/main/LICENSE for license                             
+; information.
+
+; This is a hack of the highest order.
+; The HuC6280 provides optimized (17+(6 cycles per byte)!) memory transfer
+; instructions, but they only take immediate arguments. Therefore, build
+; a subroutine in the imaginary register area and jump to it.
+
+.include "imag.inc"
+
+.global pce_memop
+.global pce_rmemcpy
+.global pce_memmove
+.global pce_memset
+
+    .section .text.pce_rmemcpy, "ax", @progbits
+; __rc2-__rc3 dest
+; __rc4-__rc5 source
+;   X  -  A   length
+pce_rmemcpy:
+    sta __rc8
+    stx __rc9
+    sec
+    sbc #1
+    sta __rc6
+    sax
+    sbc #0
+    sta __rc7
+    sax
+    clc
+    adc __rc4
+    sta __rc4
+    sax
+    adc __rc5
+    sta __rc5
+    lda __rc2
+    clc
+    adc __rc6
+    sta __rc6
+    lda __rc3
+    adc __rc7
+    sta __rc7
+    lda #$C3 ; TDD
+    bra __pce_memop_inner
+
+    .section .text.pce_memmove, "ax", @progbits
+; __rc2-__rc3 dest
+; __rc4-__rc5 source
+;   X  -  A   length
+pce_memmove:
+    ldy __rc4
+    cpy __rc2
+    bcc pce_rmemcpy
+    ldy __rc5
+    cpy __rc3
+    bcc pce_rmemcpy
+    bra pce_memcpy
+
+    .section .text.pce_memop, "ax", @progbits
```

Comment:
There's a little choice point here: bunding pce_memop and __pce_memop_inner together like this makes them a unit for linker GC; accordingly the 14 byte prefix will always be included, even if never used. You could instead break them into separate sections and have a bra between them; this would add an extra instruction to pce_memop, but it would save those 14 bytes in the case that only e.g. pce_memmove and friends are used, which seems likely.

Reply:
The only situation in which this would help is if only `pce_memset` is used; all the other.

Besides, these functions need a bit of a rewrite anyway - studying Elmer's notes reveals that, contrary to my belief, HuC block move instructions *do* stall interrupts, making them infeasible as a general-case memcpy replacement. Might be best to simply drop them (with the exception of pce_memop) from the initial release, which is what I'm going to do.